lib/vrl/compiler/src/expression/op.rs

```diff
@@ -93,28 +93,37 @@ impl Expression for Op {
         use ast::Opcode::*;
         use Value::*;
 
-        let lhs = self.lhs.resolve(ctx);
-        let mut rhs = || self.rhs.resolve(ctx);
+        if let Err = self.opcode {
+            return self.lhs.resolve(ctx).or_else(|_| self.rhs.resolve(ctx));
+        } else if let Or = self.opcode {
+            return self
+                .lhs
+                .resolve(ctx)?
+                .try_or(|| self.rhs.resolve(ctx))
+                .map_err(Into::into);
+        };
+
+        let lhs = self.lhs.resolve(ctx)?;
+        let rhs = self.rhs.resolve(ctx)?;
 
         match self.opcode {
-            Mul => lhs?.try_mul(rhs()?),
-            Div => lhs?.try_div(rhs()?),
-            Add => lhs?.try_add(rhs()?),
-            Sub => lhs?.try_sub(rhs()?),
-            Rem => lhs?.try_rem(rhs()?),
-            Or => lhs?.try_or(rhs),
-            And => match lhs? {
+            Mul => lhs.try_mul(rhs),
+            Div => lhs.try_div(rhs),
+            Add => lhs.try_add(rhs),
+            Sub => lhs.try_sub(rhs),
+            Rem => lhs.try_rem(rhs),
+            And => match lhs {
```

Comment:
`And` also needs to short-circuit if the first value is false

Reply:
Good catch!

Reply:
Resolved in [60ca3a1](https://github.com/vectordotdev/vector/pull/12681/commits/60ca3a1de49851670fd063048b7ae4e21652d818).